## D17 · 异常处理与错误码

### 1. 目标

建立一套**全链路、标准化**的异常处理机制和错误码规范。该规范旨在确保当系统发生预期或意外的错误时，能够被优雅地捕获、传递和处理，并向用户提供清晰、有用的反馈，同时为监控和告警提供结构化的数据。

### 2. 错误处理原则

*   **尽早失败 (Fail Fast)**: 在请求的入口处（如 API Gateway, Controller）进行严格的参数校验，对于无效请求应尽早拒绝，避免其进入业务逻辑深处。
*   **向上冒泡**: 后端业务逻辑中的异常应被捕获，并包装成标准化的业务错误，向上层抛出，最终由全局异常处理器统一处理。
*   **用户友好**: 返回给前端的错误信息，应包含**机器可读的错误码**和**人类可读的提示语**。敏感的内部错误细节（如堆栈跟踪）**绝不能**泄露给客户端。
*   **可追溯性**: 每一次错误都必须被记录到日志中，并包含一个唯一的 `traceId`，以便于事后排查。

### 3. 统一错误返回结构

所有 REST API 和 WebSocket 的 `error` 类型的响应，都应遵循统一的结构。

```json
{
  "ok": false, // 明确表示请求失败
  "error": {
    "code": "INVALID_PHASE", // 机器可读的错误码
    "message": "Action is not allowed in the current game phase.", // 对开发人员友好的调试信息
    "hint": "Please wait for the next phase to perform this action.", // 可直接向用户展示的提示语
    "traceId": "trace-uuid-abc-123" // 用于追踪的 ID
  }
}
```

### 4. 错误码枚举

错误码应按类别进行划分，便于管理和理解。

| 类别 (`Category`) | 错误码 (`Code`)          | HTTP 状态 | 描述                                                         |
| ----------------- | ------------------------ | --------- | ------------------------------------------------------------ |
| **通用 (General)**  | `BAD_REQUEST`            | 400       | 请求参数格式错误或缺失。                                     |
|                   | `UNAUTHORIZED`           | 401       | 未提供有效的身份认证令牌。                                   |
|                   | `FORBIDDEN`              | 403       | 已认证，但无权访问该资源。                                   |
|                   | `NOT_FOUND`              | 404       | 请求的资源不存在。                                           |
|                   | `CONFLICT`               | 409       | 请求与服务器当前状态冲突。                                   |
|                   | `RATE_LIMITED`           | 429       | 请求频率过高。                                               |
|                   | `INTERNAL_SERVER_ERROR`  | 500       | 服务器内部发生未知错误。                                     |
| **房间 (Room)**     | `ROOM_NOT_FOUND`         | 404       | 房间不存在或已关闭。                                         |
|                   | `ROOM_FULL`              | 409       | 房间已满员。                                                 |
|                   | `SEAT_TAKEN`             | 409       | 指定的座位已被占用。                                         |
|                   | `GAME_ALREADY_STARTED`   | 409       | 游戏已开始，无法加入或修改设置。                             |
| **游戏逻辑 (Game)** | `INVALID_PHASE`          | 409       | 当前游戏阶段不允许该操作。                                   |
|                   | `NOT_YOUR_TURN`          | 403       | 不是该玩家的行动回合。                                       |
|                   | `TARGET_INVALID`         | 400       | 选择的目标无效（如已死亡或不存在）。                         |
|                   | `SKILL_NOT_AVAILABLE`    | 403       | 角色技能不可用（如已使用、冷却中或条件不满足）。             |
|                   | `CONTENT_REJECTED`       | 400       | 发言内容因包含违规信息被拒绝。                               |

### 5. 前端处理策略

前端需要一个全局的错误处理器（例如，axios 的 interceptor）来统一处理来自后端的错误。

*   **可重试错误**: 对于 `RATE_LIMITED` 或网络超时等临时性错误，UI 可以提供一个“重试”按钮，并实现指数退避的重试逻辑。
*   **业务逻辑错误**: 对于 `INVALID_PHASE`, `NOT_YOUR_TURN` 等明确的业务错误，UI 应禁用相关操作按钮，并根据返回的 `hint` 字段向用户显示友好的提示（Toast 或 Alert）。
*   **致命错误**: 对于 `UNAUTHORIZED` 或 `INTERNAL_SERVER_ERROR` 等严重错误，应引导用户刷新页面或返回大厅，并可能需要清除本地的会话状态。
*   **本地化**: `hint` 字段的内容可以是一个翻译键（如 `errors.invalidPhaseHint`），由前端的 i18n 模块（参见 D11）进行本地化展示。

### 6. 监控与告警

*   **日志聚合**: 所有错误日志都应被集中上报到日志系统（如 ELK/Loki）。
*   **仪表板**: 在 Grafana 中创建一个错误监控仪表板，展示：
    *   HTTP 4xx/5xx 错误率的变化趋势。
    *   Top 10 最常见的错误码。
*   **告警**: 配置告警规则，当以下情况发生时，通过 PagerDuty 或 Slack 通知开发团队：
    *   5xx 错误率在 5 分钟内激增超过 1%。
    *   某个特定的业务错误码（如 `SKILL_NOT_AVAILABLE`）的出现频率异常升高，可能预示着逻辑 bug。

### 7. 验收标准

*   **错误码覆盖率**: 后端所有可预见的业务异常，都必须映射到一个已定义的、唯一的错误码。
*   **UI 健壮性**: 通过在前端注入模拟的错误响应，验证 UI 能够优雅地处理各类错误，不会出现崩溃或白屏。
*   **日志可追溯性**: 随机抽取一次生产环境中的失败请求，验证能够通过 `traceId` 在日志系统中找到完整的错误信息和上下文。
*   **告警有效性**: 手动触发一次 5xx 错误，验证告警能够被正确触发并发送到指定的通知渠道。